[AMDGPU] Account for existing SDWA selections

frederik-h opened the pull request:
The si-peephole-sdwa pass adjusts the selections on sdwa instructions to the selections on their operands during its conversions. For instance, if an instruction selects `BYTE_0` and its operand selects `WORD_1`, the combined selection should be `BYTE_2`, i.e. "`BYTE_0` of `WORD_1`". The existing implementation does not always handle this correctly in some complex situations with instructions across different basic blocks as demonstrated by the test cases included in this PR.

This PR adds an additional selection combination step to the conversion to fix this issue. It reverts the changes made by PR #123942 which had disabled the conversion of preexisting SDWA instructions completely as a quick fix.

jrbyrnes committed:
[AMDGPU] Account for existing SDWA selections

    Change-Id: I3e1cf6042f069e8dffe9dd5b4654288111f7b1bf

frederik-h committed:
[AMDGPU] Correct transformation and simplify combineSdwaSel

- Remove redundant "if".
- Replace arithmetic on SdwaSel type

The case distinction seems clearer and removes a mishandled case:
Since (SdwaSel)((unsigned)WORD_0 + 2) == DWORD,
the existing code led to the transformation:
     WORD_0 Sel (WORD_1 Sel (%X)) -> DWORD Sel (%X)
The correct transformation should be:
    WORD_0 Sel (WORD_1 Sel (%X)) -> WORD_1 Sel (%X)

frederik-h committed:
[AMDGPU] Change formatting of combineSdwaSel

frederik-h committed:
[AMDGPU] Remove dead branch from SIPeepholeSDWA::convertToSDWA

frederik-h committed:
[AMDGPU] Extract SDWA instruction creation from convertToSDWA

frederik-h committed:
[AMDGPU] Unify loops in SIPeepholeSDWA::convertToSDWA

There are  two loops that invoke the conversion on the operands
of the input instruction, one for the case where the instruction
is already an SDWA instruction and one for the case where it isn't.
The loops are almost the same.

Fuse those loops into a single loop.

frederik-h committed:
[AMDGPU] Invert if statement in SIPeepholeSDWA::convertToSDWA

frederik-h committed:
[AMDGPU] Rename "Combine" to "CombineSelections" in SIPeepholeSDWA

frederik-h committed:
[AMDGPU] Change combineSdwaSel to use optional return type

frederik-h committed:
[AMDGPU] Add regression test for invalid SDWA selection handling

frederik-h committed:
[AMDGPU] clang-format changes to SIPeepholeSDWA

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel.mir:
```diff
@@ -0,0 +1,124 @@
+    S_ENDPGM 0
+
+...
+
```
There's a lot going on in this MIR, do we really need control flow? I'd also like an end to end IR test for any miscompiles

frederik-h replied:
> There's a lot going on in this MIR, do we really need control flow? I'd also like an end to end IR test for any miscompiles

I have simplified the test in the alternative PR. Does this look better: https://github.com/llvm/llvm-project/blob/02279b4168f23a1d35cb1217c614fb0664b7ba1a/llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel.mir? I needed at least one jump to reproduce the incorrect selection, i.e. the erroneous sdwa selections changed once I moved the instructions from bb.2 into bb.0 - I have not yet understood why this is happening.

frederik-h replied:
> There's a lot going on in this MIR, do we really need control flow? I'd also like an end to end IR test for any miscompiles
I think the control flow is necessary to reproduce the issue. The `SIPeepholeSDWA::run(MachineFunction &MF)` function changes the instructions within a single basic block until a fixed point is reached, but the basic blocks are only visited once, i.e. the loop structure looks like this:
```cpp
// Find all SDWA operands in MF.
bool Ret = false;
for (MachineBasicBlock &MBB : MF) {
    bool Changed = false;
    do {
        matchSDWAOperands(MBB);
        
        [...]
    
     } while (Changed);
}
```

Here is minimized reproducer for the incorrect sdwa selections which was derived from original program that produced a wrong output because of this issue:   
```cpp
const int n = 4;

__device__ void u(int *ap, uint8_t (&at)[n]) {
  for (int i = n - 1; i; --i) {
    if (at[i - 1])
      ap[at[i]] = i;
  }
}

struct al {
  static __global__ void ar(uint8_t *aa) {
    uint8_t at[n];
    __shared__ int *ap;

    // This used to be a loop but unrolling it
    // preserves the problematic sdwa instructions:
    // for (int i = 0; i < n; i++)
    //    at[i] = aa[threadIdx.x + i];
    at[0] = aa[threadIdx.x + 0];
    at[1] = aa[threadIdx.x + 1];
    at[2] = aa[threadIdx.x + 2];
    at[3] = aa[threadIdx.x + 3];

    // Some loop and conditional statement
    // with a body that accesses ap seems to be
    // necessary to reproduce the issue
    // since the sdwa instructions disappear otherwise.
    // For instance, this would work as well:
    //     for (int i = 0; i < 2; i++)
    //         if (!aa)
    //           ap = (int*)nullptr;
    for (int i = 0; i < n; i++) {
      if (threadIdx.x == i)
        ap[i] = 0;
    }

    // If the call is inlined, sdwa instructions are still used, but the sdwa selection combination
    // has no effect.
    u(ap, at);
  }
};

```

The difference in the instructions between the baseline and the version produced by the code from this PR is:
```diff
<       v_lshlrev_b32_sdwa v6, v5, v1 dst_sel:DWORD dst_unused:UNUSED_PAD src0_sel:DWORD src1_sel:WORD_1
---
>       v_lshlrev_b32_sdwa v6, v5, v1 dst_sel:DWORD dst_unused:UNUSED_PAD src0_sel:DWORD src1_sel:BYTE_2
```

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -102,12 +105,47 @@ class SDWAOperand {
   MachineOperand *getReplacedOperand() const { return Replaced; }
   MachineInstr *getParentInst() const { return Target->getParent(); }
 
+  /// Fold a \p FoldedOp SDWA selection into an \p ExistingOp existing SDWA
+  /// selection. If the selections are compatible, return the combined
+  /// selection, otherwise return a nullopt. For example, if we have existing
+  /// BYTE_0 Sel and are attempting to fold WORD_1 Sel:
+  ///     BYTE_0 Sel (WORD_1 Sel (%X)) -> BYTE_2 Sel (%X)
```
So this patch is implementing folding of SDWA into SDWA? In general that seems like something we should avoid the need for up-front. In terms of avoiding miscompiles, can we do a simpler patch which just avoids the miscompile by skipping SDWA+SDWA cases first?

frederik-h replied:
> So this patch is implementing folding of SDWA into SDWA?

I would not put it this way. The pass does already handle existing SDWA instructions, see e.g. https://github.com/llvm/llvm-project/blob/48d0ef1a07993139e1acf65910704255443103a5/llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp#L1032.  It does not account for the selections correctly. This PR changes the handling to skip the creation of a new SDWA instruction and to combine the selections. It's more of a bugfix and not a new feature.

> In general that seems like something we should avoid the need for up-front.

But does this mean that any handling of pre-existing SDWA instructions should be removed from [here](https://github.com/llvm/llvm-project/blob/48d0ef1a07993139e1acf65910704255443103a5/llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp#L1032), e.g. by inserting a `assert (!TII->isSDWA(Opcode))`  (or using a more graceful exit)? 

> In terms of avoiding miscompiles, can we do a simpler patch which just avoids the miscompile by skipping SDWA+SDWA cases first?

I will consider this and discuss it with @jrbyrnes, but my impression is that SIPeepholeSDWA should be enabled to handle this case correctly.

jrbyrnes replied:
> > In terms of avoiding miscompiles, can we do a simpler patch which just avoids the miscompile by skipping SDWA+SDWA cases first?

> I will consider this and discuss it with @jrbyrnes, but my impression is that SIPeepholeSDWA should be enabled to handle this case correctly.

I agree with both points; I think we should implement a simple patch to disallow SDWA->SDWA cases while working on the PR to actually handle them. I think such foldings won't be too uncommon for 8 bit code.

arsenm replied:
I suppose this is a plausible pattern due to the DAG cross block optimization issue

frederik-h replied:
> > > In terms of avoiding miscompiles, can we do a simpler patch which just avoids the miscompile by skipping SDWA+SDWA cases first?
> 
> > I will consider this and discuss it with @jrbyrnes, but my impression is that SIPeepholeSDWA should be enabled to handle this case correctly.
> 
> I agree with both points; I think we should implement a simple patch to disallow SDWA->SDWA cases while working on the PR to actually handle them. I think such foldings won't be too uncommon for 8 bit code.

I have created an alternative draft PR which tries to disable the handling of preexisting SDWA instructions in the pass:  #123942. @jrbyrnes @arsenm Does this look better to you as a first step or do we need to "disallow SDWA->SDWA" in a more selective way? In any case, I have also simplified the mir test in the other PR. The end-to-end test is still missing.

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into SIPeepholeSDWA-CombineSelections

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Reenable existing SDWA instruction handling

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Stop using CombineSelections in convertToSDWA

The flag is not necessary since the relevant instructions
can be detected by looking at the SrcSel->getImm().

frederik-h committed:
[AMDGPU] SIPeepholeSDWA.cpp: Simplify combineSdwaSel uses

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Change arg names and comments

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into SIPeepholeSDWA-CombineSelections

frederik-h committed:
[AMDGPU] Use default check prefix in sdwa-peephole-instr-combine-sel.mir

frederik-h commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1172,6 +1209,21 @@ bool SIPeepholeSDWA::convertToSDWA(MachineInstr &MI,
+    SDWAInst = MI.getParent()->getParent()->CloneMachineInstr(&MI);
+    MI.getParent()->insert(MI.getIterator(), SDWAInst);
+  } else {
+    SDWAInst = createSDWAVersion(MI);
```
The changes made to this function (i.e. extraction of `createSDWAVersion` and the inversion of the conditional statement below) are remainders of an earlier version of this PR and not essential. I would still suggest to keep them as I believe that they improve readability.

frederik-h committed:
Revert unintended reformatting

jrbyrnes commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -108,6 +111,42 @@ class SDWAOperand {
+  /// otherwise return a nullopt. For example, if we have ExistingSel
+  /// = BYTE_0 Sel and FoldedSel WORD_1 Sel:
+  ///     BYTE_0 Sel (WORD_1 Sel (%X)) -> BYTE_2 Sel (%X)
+  std::optional<SdwaSel> combineSdwaSel(SdwaSel ExistingSel,
```
Can we disallow the bad combos when checking if its possible to convert the op / MI ? This allows for early exit + removal of the std::optional

frederik-h replied:
We could split of an `areSelectionsCompatible` predicate from this function and use it to check whether or not this function will be successful before we call it instead of using the `std::optional`. But considering the way in which `SDWASrcOperand::convertToSDWA` determines `SrcSel`, I don't see how we could move this check much above the current place at which we invoke `combineSdwaSel` in this function. Perhaps placing it above the `V_MAC` handling would work. And I am not sure if splitting the function would really improve readability. Here's how the function would roughly look like (still using the `std::optional` which I had not yet removed): 
```
diff --git a/llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp b/llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp
index 39fe65e7c375..d87072ea88eb 100644
--- a/llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp
+++ b/llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp
@@ -111,31 +111,33 @@ public:
   MachineOperand *getReplacedOperand() const { return Replaced; }
   MachineInstr *getParentInst() const { return Target->getParent(); }

+  bool compatibleSelections(SdwaSel ExistingSel, SdwaSel OperandSel) {
+    return ExistingSel == SdwaSel::DWORD || OperandSel == ExistingSel ||
+           (ExistingSel != SdwaSel::WORD_1 && ExistingSel != SdwaSel::BYTE_2 &&
+            ExistingSel != SdwaSel::BYTE_3 &&
+            (OperandSel == SdwaSel::WORD_0 || OperandSel == SdwaSel::WORD_1));
+  }
+
   /// Combine an SDWA instruction's existing SDWA selection \p
   /// ExistingSel with the SDWA selection \p OpSel of its operand. If
   /// the selections are compatible, return the combined selection,
   /// otherwise return a nullopt. For example, if we have ExistingSel
-  /// = BYTE_0 Sel and FoldedSel WORD_1 Sel:
+  /// = BYTE_0 Sel and OperandSel WORD_1 Sel:
   ///     BYTE_0 Sel (WORD_1 Sel (%X)) -> BYTE_2 Sel (%X)
   std::optional<SdwaSel> combineSdwaSel(SdwaSel ExistingSel,
                                         SdwaSel OperandSel) {
-    if (ExistingSel == SdwaSel::DWORD)
-      return OperandSel;
-
-    if (OperandSel == SdwaSel::DWORD)
-      return ExistingSel;
-
-    if (ExistingSel == SdwaSel::WORD_1 || ExistingSel == SdwaSel::BYTE_2 ||
-        ExistingSel == SdwaSel::BYTE_3)
+    if (!compatibleSelections(ExistingSel, OperandSel))
       return {};

-    if (ExistingSel == OperandSel)
-      return ExistingSel;
+    if (ExistingSel == SdwaSel::DWORD)
+      return OperandSel;

-    if (OperandSel == SdwaSel::WORD_0)
+    if (OperandSel == ExistingSel ||
+       OperandSel == SdwaSel::DWORD ||
+       OperandSel == SdwaSel::WORD_0)
       return ExistingSel;

-    if (OperandSel == SdwaSel::WORD_1) {
+        if (OperandSel == SdwaSel::WORD_1) {
       if (ExistingSel == SdwaSel::BYTE_0)
         return SdwaSel::BYTE_2;
       if (ExistingSel == SdwaSel::BYTE_1)
@@ -143,8 +145,7 @@ public:
       if (ExistingSel == SdwaSel::WORD_0)
         return SdwaSel::WORD_1;
     }
-
-    return {};
+    llvm_unreachable("Unreachable");
   }

   MachineRegisterInfo *getMRI() const {
``` 

Do you think this is preferable to the current approach?

jrbyrnes replied:
I think that implementation doesn't really have any benefit over the current proposed one.

But in `potentialToConvert` , why can't we check the combo of the current operands sdwa sel and the corresponding sdwa sel of useMI if it is sdwa?

I don't understand why

>  the way in which SDWASrcOperand::convertToSDWA determines SrcSel

 blocks this ?

frederik-h replied:
> I think that implementation doesn't really have any benefit over the current proposed one.
> 
> But in `potentialToConvert` , why can't we blacklist the bad combos of the current operands sdwa sel and the corresponding sdwa sel of useMI if it is sdwa?
> 

I did not really see how it would fit in exactly at this place. But I have now changed the way in which `matchSDWAOperand` creates the SDWAOperand instances to incorporate the selection compatibility check. This seems like the earliest possible place to handle this.

frederik-h replied:
@arsenm  @jrbyrnes Working on the test cases that @arsenm requested, I realized that this approach does not really work as I thought. I'll work a bit more on the tests tomorrow and see how to fix that.

frederik-h replied:
> @arsenm @jrbyrnes Working on the test cases that @arsenm requested, I realized that this approach does not really work as I thought. I'll work a bit more on the tests tomorrow and see how to fix that.

I got confused by the role of the different instructions when adding the early checks for compatibility of the selections. I have now moved the checks to the places which modify the `PotentialMatches` which (as @jrbyrnes essentially suggested initially) is the right place to do it.

jrbyrnes commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1172,6 +1209,21 @@ bool SIPeepholeSDWA::convertToSDWA(MachineInstr &MI,
+
+  MachineInstr *SDWAInst;
+  if (TII->isSDWA(MI.getOpcode())) {
+    SDWAInst = MI.getParent()->getParent()->CloneMachineInstr(&MI);
```
Do we really need the clone? Maybe we can handle code removal just for the case with no pre-existing sdwa

frederik-h replied:
I don't think we can avoid the clone. I actually gave it a try first - using `SDWAInst = &MI` if MI is already an SDWAInst and making the code removal conditioned on `SDWAInst != &MI`. This made, for instance, the "MC/AMDGPU/gfx10_asm_vopc_sdwa.s." test hang. It makes sense that we need to operate on a copy if you look at the `convertToSDWA` implementations for the different Operand types. They may change the MI. If the conversion fails for one of them, we need to able to revoke those changes. The base version of this code also always operates on a copy which is created in the code that has been moved into `createSDWAVersion` by this PR.

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Verify compatibility of selections earlier

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Adjust comments and variable names

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Add comment answering a review question

frederik-h committed:
clang-format changes

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -164,12 +164,34 @@ class SDWASrcOperand : public SDWAOperand {
+    }
+
+    return std::unique_ptr<SDWAOperand>(new SDWASrcOperand(
+        TargetOp, ReplacedOp, SrcSel_, Abs_, Neg_, Sext_));
```
make_unique?

I don't know why we need dynamic allocation of these. We should probably fix doing this separately

frederik-h replied:
`make_unique` does not work here because I made the constructor `protected`. I could change it but I think it is preferable to keep it this way because this should be the only way to create instances of this class.

frederik-h replied:
> I don't know why we need dynamic allocation of these. We should probably fix doing this separately

It is necessary because of the use of the abstract base class `SDWAOperand`. The closed class hierarchy could easily be replaced by a tagged union if you consider this beneficial.

frederik-h replied:
Resolving this conversation as it is no longer relevant to the current implementation.

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel.mir:
```diff
@@ -1,56 +1,124 @@
+  bb.10:
+    S_ENDPGM 0
+
 ...
```
Can we get dedicated tests for all of the mergable combinations? If you directly write the MIR with pre-folded sdwa, you shouldn't need so much MIR. IR tests would be nice but more difficult

frederik-h replied:
I'll look into this. This would clearly help in the future.

frederik-h replied:
I have now added a test that demonstrates all different selection combinations.

frederik-h commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -88,6 +88,42 @@ class SIPeepholeSDWALegacy : public MachineFunctionPass {
+/// compatible.
+/// For example, if we have Sel = BYTE_0 Sel and OperandSel = WORD_1:
+///     BYTE_0 Sel (WORD_1 Sel (%X)) -> BYTE_2 Sel (%X)
+SdwaSel combineSdwaSel(SdwaSel Sel, SdwaSel OperandSel) {
```
Did you see that this function is contained in an anonymous namespace? Should I add `static` nevertheless?

frederik-h replied:
I have now changed it to `static` and moved it out of the anonymous namespace as per the LLVM coding standards.

frederik-h commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -164,12 +164,34 @@ class SDWASrcOperand : public SDWAOperand {
+        const MachineOperand *NamedOp = TII->getNamedOperand(MI, SelOpName);
+        if (NamedOp && !compatibleSelections(
+                           static_cast<SdwaSel>(NamedOp->getImm()), SrcSel_))
+          return std::unique_ptr<SDWAOperand>(nullptr);
```
That's only possible if I change the other `return` statement to return the same type. Since this (i.e. always returning `std::unique_ptr<SDWASrcOperand>`) makes more sense anyway, I'll change it.

frederik-h committed:
Use consistent/more specific return type for SDWA{Src,Dst}Operand factory

frederik-h committed:
fixup! Use consistent/more specific return type for SDWA{Src,Dst}Operand factory

frederik-h committed:
Merge "compatibleSelections" function back into "combineSdwaSel"

frederik-h committed:
Add comprehensive test for source selection combinations

frederik-h committed:
Revert introduction of SDWA{Dst,Src}Operand::create

The place where this was used is not the right place
for performing the combineSdwaSel check.

frederik-h committed:
Fix combineSdwaSel handling of Sel == OperandSel case

frederik-h committed:
Add new early check for combinable selections

frederik-h committed:
clang-format changes

frederik-h committed:
Move combineSdwaSel from anon namespace and make 'static'

As per LLVM coding standards.

jrbyrnes commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -158,15 +166,16 @@ class SDWADstOperand : public SDWAOperand {
                                    const GCNSubtarget &ST,
                                    SDWAOperandsMap *PotentialMatches = nullptr) override;
   bool convertToSDWA(MachineInstr &MI, const SIInstrInfo *TII) override;
+  bool canCombineSelections(const MachineInstr &MI,
```
Should we also check this in `SDWADstOperand::potentialToConvert` ?

frederik-h replied:
It isn't strictly necessary since there is a check at the call site (i.e. in ` SIPeepholeSDWA::run`). Seeing that this can be confusing, I have now moved all uses of `canCombineSelections` into `potentialToConvert`. Consequently, I made the function `private`. We could also inline it if you prefer, but I have a slight preference for keeping it separate - especially since there are two uses in ` SDWASrcOperand::potentialToConvert`.

jrbyrnes commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -187,6 +196,8 @@ class SDWADstPreserveOperand : public SDWADstOperand {
         Preserve(PreserveOp) {}
 
   bool convertToSDWA(MachineInstr &MI, const SIInstrInfo *TII) override;
+  bool canCombineSelections(const MachineInstr &MI,
```
Not sure this is needed

frederik-h replied:
It was strictly necessary as long as the check was outside of `potentialToConvert`; see also https://github.com/llvm/llvm-project/pull/123221#discussion_r1967148059.

jrbyrnes commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1285,10 +1374,12 @@ bool SIPeepholeSDWA::run(MachineFunction &MF) {
+            Operand->potentialToConvert(TII, ST, &PotentialMatches);
+
+        if (PotentialMI && isConvertibleToSDWA(*PotentialMI, ST, TII) &&
+            Operand->canCombineSelections(*PotentialMI, TII))
```
Think this is redundant if we're already checking from `Operand->potentialToConvert` above

frederik-h replied:
Before my last commit, there was no equivalent check in `potentialToConvert` (i.e. on the return value of that function), but I have now moved the call to `canCombineSelections` there; see also https://github.com/llvm/llvm-project/pull/123221#discussion_r1967148059. The call that already existed in `SDWASrcOperand::potentialToConvert` checks that the uses of the operand can also be converted to SDWA instructions. This is necessary in addition to the check of the return value of `potentialToConvert` since the uses are directly added to `PotentialMatches` from there (see function `op_select_byte_0_instr_select_word_1` in `llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel2.mir` for an example which leads to a crash if this check is removed).

jrbyrnes commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel2.mir:
```diff
@@ -0,0 +1,796 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1030 -simplify-mir -run-pass=si-peephole-sdwa -o - %s | FileCheck %s
+
+# Test the combination of SDWA selections in si-peephole-sdwa. In each
```
Do we have tests for combining on dst?

frederik-h replied:
No. I considered this to "demonstrate all different selection combinations" as I wrote in another comment and those are the same for the source and destination selections. Do you think we need to add a dedicated test for the dst selections?

arsenm replied:
The destination case is significantly more likely to break, so should also cover those

frederik-h replied:
I had added a test for the dst_sel values that I saw being introduced in `SIPeepholeSDWA::matchSDWAOperands`, i.e. one of the values always has to come from a non-SDWA instruction and I think I cannot introudce arbitrary selection values in this way.

frederik-h committed:
Move all uses of "canCombineSelections" into "potentialToConvert"

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -372,7 +416,9 @@ MachineInstr *SDWASrcOperand::potentialToConvert(const SIInstrInfo *TII,
     return nullptr;
 
-  return PotentialMO->getParent();
+  auto parent = PotentialMO->getParent();
+
+  return canCombineSelections(*parent, TII) ? parent : nullptr;
```
```suggestion
  MachineInstr *Parent = PotentialMO->getParent();

  return canCombineSelections(*Parent, TII) ? Parent : nullptr;
```

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -451,13 +497,33 @@ bool SDWASrcOperand::convertToSDWA(MachineInstr &MI, const SIInstrInfo *TII) {
   copyRegOperand(*Src, *getTargetOperand());
   if (!IsPreserveSrc) {
-    SrcSel->setImm(getSrcSel());
+    SdwaSel ExistingSel = static_cast<SdwaSel>(SrcSel->getImm());
```
Does the machine verifier enforce that the SDWA values are a valid enum value? If not we should make sure it does

frederik-h replied:
It doesn't. I have tried to improve this in this PR #128898.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -451,13 +497,33 @@ bool SDWASrcOperand::convertToSDWA(MachineInstr &MI, const SIInstrInfo *TII) {
   if (!IsPreserveSrc) {
-    SrcSel->setImm(getSrcSel());
+    SdwaSel ExistingSel = static_cast<SdwaSel>(SrcSel->getImm());
+    SrcSel->setImm(combineSdwaSel(ExistingSel, getSrcSel()).value());
```
```suggestion
    SrcSel->setImm(*combineSdwaSel(ExistingSel, getSrcSel()));
```

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -451,13 +497,33 @@ bool SDWASrcOperand::convertToSDWA(MachineInstr &MI, const SIInstrInfo *TII) {
+  if (!TII->isSDWA(MI.getOpcode()))
+    return true;
+
+  auto canCombineSel = [&](auto SrcOpName, auto SrcSelOpName) {
```
Move this to a full helper instead of the lambda?

frederik-h replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -451,13 +497,33 @@ bool SDWASrcOperand::convertToSDWA(MachineInstr &MI, const SIInstrInfo *TII) {
+    const MachineOperand *Src = TII->getNamedOperand(MI, SrcOpName);
+    const MachineOperand *SrcSel = TII->getNamedOperand(MI, SrcSelOpName);
+    return !Src || !isSameReg(*Src, *getReplacedOperand()) ||
+           combineSdwaSel(static_cast<SdwaSel>(SrcSel->getImm()), getSrcSel())
+               .has_value();
```
```suggestion
           !combineSdwaSel(static_cast<SdwaSel>(SrcSel->getImm()), getSrcSel());
```

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -476,7 +542,8 @@ MachineInstr *SDWADstOperand::potentialToConvert(const SIInstrInfo *TII,
   }
 
-  return PotentialMO->getParent();
+  auto parent = PotentialMO->getParent();
+  return canCombineSelections(*parent, TII) ? parent : nullptr;
```
```suggestion
  MachineInstr *Parent = PotentialMO->getParent();
  return canCombineSelections(*Parent, TII) ? Parent : nullptr;
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel2.mir:
```diff
@@ -0,0 +1,796 @@
+    ; CHECK-NEXT: [[V_AND_B32_e64_:%[0-9]+]]:vgpr_32 = V_AND_B32_e64 255, [[V_LSHRREV_B32_sdwa]], implicit $exec
+    ; CHECK-NEXT: [[V_LSHRREV_B32_sdwa1:%[0-9]+]]:vgpr_32 = V_LSHRREV_B32_sdwa 0, [[V_LSHRREV_B32_sdwa]], 0, [[V_LSHRREV_B32_sdwa]], 0, 1, 0, 0, 0, implicit $exec
+    ; CHECK-NEXT: S_ENDPGM 0
+    %3:vgpr_32 = V_LSHRREV_B32_sdwa 0, $vgpr0, 0, $vgpr0, 0, 1, 0, 5, 0, implicit $exec
+    %8:vgpr_32 = V_AND_B32_e64 255,  %3, implicit $exec /* Select Byte_0 */
+    %9:vgpr_32 = V_LSHRREV_B32_sdwa 0, %3, 0, %8, 0, 1, 0, 0, 6, implicit $exec
```
Should compact register numbers. You can use -run-pass=none to do this automatically

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel2.mir:
```diff
@@ -0,0 +1,796 @@
+
+---
+name: op_select_byte0_instr_select_dword
+body:             |
+
```
```suggestion
tracksRegLiveness: true
body:             |
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel2.mir:
```diff
@@ -0,0 +1,796 @@
+    ; CHECK-NEXT: [[V_AND_B32_e64_:%[0-9]+]]:vgpr_32 = V_AND_B32_e64 255, [[V_LSHRREV_B32_sdwa]], implicit $exec
+    ; CHECK-NEXT: [[V_LSHRREV_B32_sdwa1:%[0-9]+]]:vgpr_32 = V_LSHRREV_B32_sdwa 0, [[V_LSHRREV_B32_sdwa]], 0, [[V_LSHRREV_B32_sdwa]], 0, 1, 0, 0, 0, implicit $exec
+    ; CHECK-NEXT: S_ENDPGM 0
+    %3:vgpr_32 = V_LSHRREV_B32_sdwa 0, $vgpr0, 0, $vgpr0, 0, 1, 0, 5, 0, implicit $exec
```
Why are these using physical registers? This should never appear here. Physical register uses should only occur in edge case tests involving physical registers, they shouldn't be used in a normal case

frederik-h replied:
I changed it.

frederik-h committed:
Update llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Update llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel2.mir

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Update llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Update llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Update llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
fixup! Update llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

frederik-h committed:
Extract canCombineSel helper function

frederik-h committed:
Split canCombineSel and reuse for SDWADstOperand::canCombineSelections

In contrast to SDWASrcOperand::canCombineSelections, in
SDWADstOperand::canCombineSelections we don't need to figure out
which operand we have. The same function could be used for both
but this would introduce a redundant check if the operand
is the vdst operand.

frederik-h committed:
Fix test: Don't use physical registers, compact registers

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into SIPeepholeSDWA-CombineSelections

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into SIPeepholeSDWA-CombineSelections

frederik-h committed:
Add test for destination selection

... and adjust test name of existing source selection test.

frederik-h committed:
Use right type for OpNames

frederik-h committed:
clang-format changes

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel-src.mir:
```diff
@@ -0,0 +1,1089 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1030 -simplify-mir -run-pass=si-peephole-sdwa -o - %s | FileCheck %s
+
```
I'm not sure I see the dest write cases?

frederik-h replied:
You are right, I missed to add the new file. Here it is now: https://github.com/llvm/llvm-project/pull/123221/commits/58e278fd77cc82527fd3f6ead618519283cfad9b#diff-30c0b74efc1ede6ed6f8915cdbfff782669645f27607e94a62a238ca82124624

frederik-h committed:
Add dst_sel test

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-instr-combine-sel-dst.mir:
```diff
@@ -0,0 +1,309 @@
+    %1:vgpr_32 = COPY $vgpr0
+    %2:vgpr_32 = V_LSHRREV_B32_sdwa 0, %1, 0, %1, 0, 6, 0, 0, 0, implicit $exec
+    %3:vgpr_32 = V_LSHLREV_B32_e32  16, %2,  implicit $exec /* Select WORD_1 */
+    %4:vgpr_32 = V_LSHRREV_B32_sdwa 0, %1, 0, %1, 0, 2, 0, 6, 6, implicit $exec
```
Do we really not have tied operands on the dest write cases

frederik-h replied:
@arsenm As we have now discussed elsewhere: I'll look into better understanding the destination selection case and if this code should perhaps be rejected by the MachineVerifier or this pass as follow-up work.
